Mathlib/LinearAlgebra/Multilinear/DirectSum.lean

```diff
@@ -0,0 +1,175 @@
+/-
+Copyright (c) 2024 Sophie Morel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sophie Morel
+-/
+import Mathlib.Algebra.DirectSum.Module
+import Mathlib.LinearAlgebra.Multilinear.Basic
+
+/-!
+# Multilinear maps from direct sums
+
+This file describes multilinear maps on direct sums.
+
+## Main results
+
+* `MultilinearMap.fromDirectSumEquiv` : If 'ι` is a `Fintype`, `κ i` is a family of types
+indexed by `ι` and we are given a `R`-module `M i j` for every `i : ι` and `j : κ i`, this is
+the linear equivalence between `Π p : (i : ι) → κ i, MultilinearMap R (fun i ↦ M i (p i)) M'` and
+`MultilinearMap R (fun i ↦ ⨁ j : κ i, M i j) M'`.
+-/
+
+namespace MultilinearMap
+
+open DirectSum LinearMap Function
+
+variable (R : Type*) [CommSemiring R]
+variable {ι : Type*} [Fintype ι] [DecidableEq ι]
+variable (κ : ι → Type*) [(i : ι) → DecidableEq (κ i)]
+variable {M : (i : ι) → κ i → Type*} {M' : Type*}
+variable [Π i (j : κ i), AddCommMonoid (M i j)] [AddCommMonoid M']
+variable [Π i (j : κ i), Module R (M i j)] [Module R M']
+variable [Π i (j : κ i) (x : M i j), Decidable (x ≠ 0)]
+
+private theorem fromDirectSum_aux1 (f : Π (p : Π i, κ i), MultilinearMap R (fun i ↦ M i (p i)) M')
+    (x : Π i, ⨁ (j : κ i), M i j) {p : Π i, κ i}
+    (hp : p ∉ Fintype.piFinset (fun i ↦ (x i).support)) :
+    (f p) (fun j ↦ (x j) (p j)) = 0 := by
+  simp only [Fintype.mem_piFinset, DFinsupp.mem_support_toFun, ne_eq, not_forall, not_not] at hp
+  obtain ⟨i, hi⟩ := hp
+  exact (f p).map_coord_zero i hi
+
+omit [Fintype ι] [(i : ι) → DecidableEq (κ i)]
+  [(i : ι) → (j : κ i) → (x : M i j) → Decidable (x ≠ 0)] in
+private theorem fromDirectSum_aux2 (x : Π i, ⨁ (j : κ i), M i j) (i : ι) (p : Π i, κ i)
+    (a : ⨁ (j : κ i), M i j) :
+    (fun j ↦ (update x i a j) (p j)) = update (fun j ↦ x j (p j)) i (a (p i)) := by
+  ext j
+  by_cases h : j = i
+  · rw [h]; simp only [update_same]
+  · simp only [ne_eq, h, not_false_eq_true, update_noteq]
+
+/-- Given a family indexed by `p : Π i : ι, κ i` of multilinear maps on the
+`fun i ↦ M i (p i)`, construct a multilinear map on `fun i ↦ ⨁ j : κ i, M i j`.-/
+def fromDirectSum (f : Π (p : Π i, κ i), MultilinearMap R (fun i ↦ M i (p i)) M') :
+    MultilinearMap R (fun i ↦ ⨁ j : κ i, M i j) M' where
+  toFun x := ∑ p in Fintype.piFinset (fun i ↦ (x i).support), f p (fun i ↦ x i (p i))
```

Comment:
I think there's another direct sum hiding here; I've created #17881 with the DFinsupp version.

This then becomes a sum over the support.

Reply:
Turns out this needs #17932 too. I'm working in https://github.com/leanprover-community/mathlib4/tree/eric-wieser/SM.PiTensorProduct.DirectSum on the union of these PRs.

Reply:
I've pushed this refactor,